### PR TITLE
add reusable code for blueprints , create extension points

### PR DESCRIPTION
**Pull request by tan-duong**
To-do
- testing: install messaging blueprint with install.sh
- run on ios/android
- run on Circle CI

### CODE REVIEW

File: {{cookiecutter.project_slug}}/package.json [hunk 1]
```diff
@@ -47,7 +47,8 @@
     "react-navigation-stack": "1.5.3",
     "react-redux": "7.1.1",
     "redux": "4.0.4",
-    "redux-saga": "1.0.5"
+    "redux-saga": "1.0.5",
+    "react-native-keyboard-aware-scroll-view": "^0.9.1"
```
Comment (indie-rok) [inline {{cookiecutter.project_slug}}/package.json]: Remove the ^ please

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import _ from 'lodash';
```
Comment (indie-rok) [inline {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: import find from 'loadash/find' is way faster 👍

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import users from './raw/users';
+import articles from './raw/articles';
+import notifications from './raw/notifications';
+import conversations from './raw/conversations';
+
+const populateArticles = () => {
+  articles.forEach(article => {
+    const userArticle = article;
```
Comment (indie-rok) [inline {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: we don't need this line, you could just name the above line `userArticle`

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    const userId = articles.indexOf(article) % users.length;
+    userArticle.user = _.find(users, x => x.id === userId) || users[0];
```
Comment (indie-rok) [inline {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: why not just use `find` from ES6?

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      return userComment;
+    });
+    return userArticle;
+  });
+};
+
+const populateNotifications = () => {
+  notifications.map(notification => {
+    const userNotification = notification;
+    const userId = notifications.indexOf(notification) % users.length;
```
Comment (indie-rok) [inline {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: You could use map index or uuid package to generate random ids, I find this way particularly harder to understand 🤔

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+const populateConversations = () => {
+  conversations.map(conversation => {
+    const userConversation = conversation;
+    userConversation.withUser = _.find(users, x => x.id === conversation.withUserId) || users[0];
+    return userConversation;
+  });
+};
+
+const populate = () => {
```
Comment (indie-rok) [inline {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: `generateSeeds` or something related  would be a more readable name for this function

File: {{cookiecutter.project_slug}}/src/data/index.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+import _ from 'lodash';
+import populate from './dataGenerator';
+import users from './raw/users';
+import articles from './raw/articles';
+import notifications from './raw/notifications';
+import conversations from './raw/conversations';
+import cards from './raw/cards';
+
+class DataProvider {
```
Comment (indie-rok) [inline {{cookiecutter.project_slug}}/src/data/index.js]: This is very cool 👍

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 2 above)
Comment (tan-duong) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: no need

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 3 above)
Comment (tan-duong) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: it was tested

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 4 above)
Comment (tan-duong) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: both works

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 5 above)
Comment (tan-duong) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: it's not random value

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 6 above)
Comment (tan-duong) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: no need to change the name

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 2 above)
Comment (indie-rok) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: There is need, as mentioned, `lodash/find` is faster while compiling build.

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 3 above)
Comment (indie-rok) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: I know, I tested my self too. As mentioned is a line of code that it is not necessary and makes the code harder to understand

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 4 above)
Comment (indie-rok) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: I know it works, I tested my self, find from `ES6` is native and faster to create JS build

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 6 above)
Comment (indie-rok) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: There is a need `populate` is a very generic function name that makes code harder to read and understand. (especially for future developers)

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 3 above)
Comment (indie-rok) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: I take that back, think that if you are going to modify `article` it makes sense to create this new variable 👍

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 4 above)
Comment (tan-duong) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: let's have a new PR to remove lodash

File: {{cookiecutter.project_slug}}/src/data/dataGenerator.js (hunk 2 above)
Comment (tan-duong) [reply {{cookiecutter.project_slug}}/src/data/dataGenerator.js]: let's have a new PR to remove lodash

**Pull request merged**